File: universal-adapter-java/src/main/java/com/jacekmarchwicki/universaladapter/BaseAdapterItem.java
/*
 * Copyright 2016 Jacek Marchwicki <jacek.marchwicki@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package com.jacekmarchwicki.universaladapter;


public interface BaseAdapterItem {

    /**
     * return this id if you don't have any id's
     */
    long NO_ID = -1;

    /**
     * Unique adapter id or {@link #NO_ID}
     *
     * @return adapter id or {@link #NO_ID}
     */
    long adapterId();

    /**
     * If both items are the same but can have different content
     * <p>Usually it means booth items has same id</p>
     *
     * @param item to compare
     * @return true if booth items matches
     */
    boolean matches(BaseAdapterItem item);

    /**
     * If both items has exactly same content
     * <p>Usually it means both items has same id, name and other fields</p>
     * <p>If you implemented {@link Object#equals(Object)} you can call
     * {@code this.equals(item)}</p>
     *
     * @param item to compare
     * @return true if booth items are the same
     */
    boolean same(BaseAdapterItem item);
}

File: universal-adapter/src/main/java/com/jacekmarchwicki/universaladapter/BaseAdapterItem.kt
/*
 * Copyright 2016 Jacek Marchwicki <jacek.marchwicki@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.jacekmarchwicki.universaladapter

interface BaseAdapterItem {
    /**
     * Unique adapter id or [BaseAdapterItem.NO_ID]
     *
     * @return adapter id or [BaseAdapterItem.NO_ID]
     */
    fun adapterId(): Long

    /**
     * If both items are the same but can have different content
     *
     * Usually it means that both items have the same id
     *
     * @param item to compare
     * @return true if both items matches
     */
    fun matches(item: BaseAdapterItem): Boolean

    /**
     * If both items have exactly the same content
     *
     * Usually it means both items have the same id, name and other fields
     *
     * If you implemented [Any.equals] you can call
     * `this.equals(item)`
     *
     * @param item to compare
     * @return true if booth items are the same
     */
    fun same(item: BaseAdapterItem): Boolean

    companion object {
        /**
         * return this id if you don't have any id's
         */
        const val NO_ID: Long = -1
    }
}